### PR TITLE
KEYCLOAK-11019 Initial support for lazy offline user-session loading

### PR DESCRIPTION
**Pull request by martin-kanis**
This PR includes #7722 + rebase to current master + adds model tests

### CODE REVIEW

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you for the PR. I went thru the original PR and there has been done huge amount of work. Thanks again.

Few notes:

- There was some issues in GHA 
    - model tests were cancelled after 329m. But I can see it passed in @martin-kanis's fork so I assume it is not related. 
    - cluster tests are failing
- I haven't found any documentation PR, is there a plan to send one?
- the changeSet should be updated, please see inline.

Otherwise it looks good to me.

**Comment by hmlnarik** [comment]
> model tests were cancelled after 329m. But I can see it passed in @martin-kanis's fork so I assume it is not related.

That's correct, it is an independent intermittent issue, tracked as https://issues.redhat.com/browse/KEYCLOAK-18091

File: model/jpa/src/main/resources/META-INF/jpa-changelog-13.0.0.xml [hunk 1]
```diff
@@ -87,4 +87,23 @@
         <renameColumn tableName="REALM_ATTRIBUTE" oldColumnName="VALUE_NEW" newColumnName="VALUE" columnDataType="NCLOB"/>
     </changeSet>
 
+    <changeSet author="keycloak" id="13.0.0-KEYCLOAK-11019">
```
[the remark this answers is not in the document]
Comment (martin-kanis) [reply model/jpa/src/main/resources/META-INF/jpa-changelog-13.0.0.xml]: Thanks, moved.

**Comment by martin-kanis** [comment]
The failing cluster tests were caused by a syntax error in SQL query on MySQL. The `coalesce` + `cast` functions didn't work as expected at least on MySQL in clustered environment with provided `null` values. I removed the `coalesce` from all queries because from the keycloak code we always called the queries with `null` values. 
I also removed some newly introduced methods which weren't called from anywhere. 

```
WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (ServerService Thread Pool -- 69) SQL Error: 1064, SQLState: 42000
2021-04-02T22:46:06.7960979Z AuthenticationSessionClusterTest ++ 22:45:45,473 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (ServerService Thread Pool -- 69) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'longtext), persistent0_.REALM_ID)' at line 1
2021-04-02T22:46:06.7965147Z AuthenticationSessionClusterTest ++ 22:45:45,484 FATAL [org.keycloak.services] (ServerService Thread Pool -- 69) Error during startup: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
2021-04-02T22:46:06.7969007Z AuthenticationSessionClusterTest ++ 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
2021-04-02T22:46:06.7972268Z AuthenticationSessionClusterTest ++ 	at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1575)
2021-04-02T22:46:06.7976018Z AuthenticationSessionClusterTest ++ 	at org.hibernate.query.internal.AbstractProducedQuery.getSingleResult(AbstractProducedQuery.java:1614)
2021-04-02T22:46:06.7980927Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.jpa.session.JpaUserSessionPersisterProvider.getUserSessionsCount(JpaUserSessionPersisterProvider.java:460)
2021-04-02T22:46:06.7986423Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.session.UserSessionPersisterProvider.getUserSessionsCount(UserSessionPersisterProvider.java:133)
2021-04-02T22:46:06.7993299Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.computeLoaderContext(OfflinePersistentUserSessionLoader.java:67)
2021-04-02T22:46:06.8001438Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.OfflinePersistentUserSessionLoader.computeLoaderContext(OfflinePersistentUserSessionLoader.java:37)
2021-04-02T22:46:06.8007571Z AuthenticationSessionClusterTest ++ 	at org.keycloak.models.sessions.infinispan.initializer.InfinispanCacheInitializer$2.run(InfinispanCacheInitializer.java:92)
```

**Comment by hmlnarik** [comment]
https://master-jenkins/job/universal-test-pipeline-server/406/

**Comment by martin-kanis** [comment]
https://master-jenkins/job/universal-test-pipeline-server/438/

**Review by vramik — COMMENTED** [review]
Thanks @martin-kanis! 

Th PR looks good to me. From my POV there is just documentation PR missing and we've discussed with @hmlnarik that after those changes in queries the index analysis (which was done with the original PR) should re-checked.

**Comment by hmlnarik** [comment]
Indices checked, they look good. @vramik , is the docs the only remaining concern?

**Review by vramik — APPROVED** [review]
@hmlnarik yes, the docs, are the only remaining concern, thank you for checking the indices.

**Review by hmlnarik — APPROVED** [review]
Pipelines ran successfully. Thank you for the contribution, @thomasdarimont and @martin-kanis !

**Pull request merged**